File: task5/task5.cpp
#include <iostream>
using namespace std;

int lastNum(int num, int degree) {
    if (degree == 0) return 1; // Любое число в степени 0 равно 1

    num = num % 10; // Переводим основание в единицы

    // Находим остаток от деления на 4, т.к. период длины 4 для последних цифр возведений чисел в степень
    int ostatok = degree % 4;
    if (ostatok == 0) ostatok = 4; // Корректируем для четырехстепенного цикла

    int result = 1;
    for (int i = 0; i < ostatok; ++i) {
        result = (result * num) % 10; // Находим последнюю цифру возведения числа в степень
    }

    return result;
}

int main() {
    setlocale(LC_ALL, "Russian");
    cout << "Введите число: ";
    int num;
    cin >> num;
    cout << "Введите первую степень: ";
    int degree1;
    cin >> degree1;
    cout << "Введите вторую степень: ";
    int degree2;
    cin >> degree2;

    int finalDegree = lastNum(degree1, degree2);
    int result = lastNum(num, finalDegree);

    cout << "Последняя цифра числа " << num << "^(" << degree1 << "^" << degree2 << ") равна: " << result << endl;

    return 0;
}

File: task5/task5.rs
use std::io;

fn last_num(mut num: u32, mut degree: u32) -> u32 {
    if degree == 0 {
        return 1; // Любое число в степени 0 равно 1
    }

    num %= 10; // Переводим основание в единицы

    // Находим остаток от деления на 4, т.к. период длины 4 для последних цифр возведений чисел в степень
    let ostatok = degree % 4;
    let mut ostatok = if ostatok == 0 { 4 } else { ostatok }; // Корректируем для четырехстепенного цикла

    let mut result = 1;
    while ostatok > 0 {
        result = (result * num) % 10; // Находим последнюю цифру возведения числа в степень
        ostatok -= 1;
    }

    result
}

fn main() {
    println!("Введите число: ");
    let mut num = String::new();
    io::stdin().read_line(&mut num).expect("Не удалось прочитать строку");
    let num: u32 = num.trim().parse().expect("Введите число");

    println!("Введите первую степень: ");
    let mut degree1 = String::new();
    io::stdin().read_line(&mut degree1).expect("Не удалось прочитать строку");
    let degree1: u32 = degree1.trim().parse().expect("Введите число");

    println!("Введите вторую степень: ");
    let mut degree2 = String::new();
    io::stdin().read_line(&mut degree2).expect("Не удалось прочитать строку");
    let degree2: u32 = degree2.trim().parse().expect("Введите число");

    let finalDegree = last_num(degree1, degree2);
    let result = last_num(num, finalDegree);

    println!(
        "Последняя цифра числа { }^({ }^{ }) равна: { }",
        num, degree1, degree2, result
    );
}
